File: backend/app/extractors/manufacturer.py
"""Extract manufacturer / packer / importer address per Rule 6(1)(a) + Rule 10."""

from __future__ import annotations

import re

from app.domain import ExtractedField, ImageMeta, OCRWord
from app.extractors.base import avg_confidence, merge_bboxes

ROLE_KEYWORDS = re.compile(
    r"\b(?:mfg|mfd|pkd|mktd|mfg\.?|mfd\.?|pkd\.?|mktd\.?|"
    r"(?:mfg|mfd)\.?\s*(?:&|and)\s*pkd\.?\s*(?:by)?:?|"
    r"manufactured\s+by:?|packed\s+by:?|pkd\.?\s*by:?|mktd\.?\s*by:?|"
    r"imported\s+by:?|marketed\s+by:?|manufactured\s+for:?|"
    r"imported\s+and\s+packed\s+by:?|manufacturer:?|packer:?)\b",
    re.IGNORECASE,
)
ADDRESS_HINT = re.compile(
    r"\b(?:pvt|ltd|limited|private|company|co\.|india|industries|foods|plot|road|street|"
    r"sector|phase|marg|nagar|colony|estate|complex|tel|phone|email|pin|"
    r"taluk|dist|district|village|khasra|survey|post|state|regd|office)\b",
    re.IGNORECASE,
)


def _group_into_lines(words: list[OCRWord], y_tolerance: float = 10.0) -> list[list[OCRWord]]:
    if not words:
        return []
    sorted_w = sorted(words, key=lambda w: (w.bbox[1], w.bbox[0]))
    lines = [[sorted_w[0]]]
    for w in sorted_w[1:]:
        if abs(w.bbox[1] - lines[-1][-1].bbox[1]) <= y_tolerance:
            lines[-1].append(w)
        else:
            lines.append([w])
    return lines


def _line_text(line: list[OCRWord]) -> str:
    return " ".join(w.text for w in line)


def _line_is_address(line: list[OCRWord], pin_regex: re.Pattern[str]) -> bool:
    return bool(pin_regex.search(_line_text(line)) or ADDRESS_HINT.search(_line_text(line)))


def extract_manufacturer_address(
    ocr_words: list[OCRWord], image_meta: ImageMeta, pin_regex: str
) -> ExtractedField | None:
    pin_re = re.compile(pin_regex)
    is_normalized = bool(
        ocr_words
        and all(
            w.bbox[0] <= 1.0 and w.bbox[1] <= 1.0 and w.bbox[2] <= 1.0 and w.bbox[3] <= 1.0
            for w in ocr_words
        )
    )
    if is_normalized:
        y_tol = (
            max(10.0, float(image_meta.height // 100)) / image_meta.height
            if image_meta.height > 0
            else 0.02
        )
    else:
        y_tol = float(max(10, image_meta.height // 100))
    lines = _group_into_lines(ocr_words, y_tol)
    start_idx = next(
        (i for i, line in enumerate(lines) if ROLE_KEYWORDS.search(_line_text(line))), None
    )
    if start_idx is None:
        return ExtractedField("manufacturer_address", None, None, 0.0, [])
    block_lines = lines[start_idx : start_idx + 1]
    has_pin = any(pin_re.search(w.text) for w in block_lines[0])
    if not has_pin:
        for j in range(start_idx + 1, min(start_idx + 7, len(lines))):
            if _line_is_address(lines[j], pin_re):
                block_lines.append(lines[j])
                if any(pin_re.search(w.text) for w in lines[j]):
                    break
            elif len(block_lines) >= 2:
                break
    block_words = [w for line in block_lines for w in line]
    pin_bbox = next((w.bbox for w in block_words if pin_re.search(w.text)), None)
    return ExtractedField(
        "manufacturer_address",
        " ".join(_line_text(line) for line in block_lines) if pin_bbox else None,
        merge_bboxes(block_words),
        avg_confidence(block_words) if pin_bbox else 0.0,
        [pin_bbox] if pin_bbox else [],
    )

File: frontend/lib/rules/extractors/manufacturer.ts
import type { ExtractedField, ImageMeta, OCRWord } from '../domain';
import { avgConfidence, mergeBboxes } from './base';

export const ROLE_KEYWORDS =
  /\b(?:mfg|mfd|pkd|mktd|mfg\.?|mfd\.?|pkd\.?|mktd\.?|(?:mfg|mfd)\.?\s*(?:&|and)\s*pkd\.?\s*(?:by)?:?|manufactured\s+by:?|packed\s+by:?|pkd\.?\s*by:?|mktd\.?\s*by:?|imported\s+by:?|marketed\s+by:?|manufactured\s+for:?|imported\s+and\s+packed\s+by:?|manufacturer:?|packer:?)\b/i;

export const ADDRESS_HINT =
  /\b(?:pvt|ltd|limited|private|company|co\.|india|industries|foods|plot|road|street|sector|phase|marg|nagar|colony|estate|complex|tel|phone|email|pin|taluk|dist|district|village|khasra|survey|post|state|regd|office)\b/i;

function groupIntoLines(words: OCRWord[], yTolerance = 10.0): OCRWord[][] {
  if (words.length === 0) return [];
  const sortedW = words
    .slice()
    .sort((a, b) => a.bbox[1] - b.bbox[1] || a.bbox[0] - b.bbox[0]);
  const lines: OCRWord[][] = [[sortedW[0]]];
  for (let i = 1; i < sortedW.length; i++) {
    const w = sortedW[i];
    const prevLine = lines[lines.length - 1];
    const lastInPrevLine = prevLine[prevLine.length - 1];
    if (Math.abs(w.bbox[1] - lastInPrevLine.bbox[1]) <= yTolerance) {
      prevLine.push(w);
    } else {
      lines.push([w]);
    }
  }
  return lines;
}

function lineText(line: OCRWord[]): string {
  return line.map((w) => w.text).join(' ');
}

function lineIsAddress(line: OCRWord[], pinRegex: RegExp): boolean {
  const text = lineText(line);
  return pinRegex.test(text) || ADDRESS_HINT.test(text);
}

export function extractManufacturerAddress(
  ocrWords: OCRWord[],
  imageMeta: ImageMeta,
  pinRegexStr: string
): ExtractedField {
  const pinRe = new RegExp(pinRegexStr);
  const isNormalized =
    ocrWords.length > 0 &&
    ocrWords.every(
      (w) =>
        w.bbox[0] <= 1.0 &&
        w.bbox[1] <= 1.0 &&
        w.bbox[2] <= 1.0 &&
        w.bbox[3] <= 1.0
    );

  let yTol: number;
  if (isNormalized) {
    yTol =
      imageMeta.height > 0
        ? Math.max(10.0, Math.floor(imageMeta.height / 100)) / imageMeta.height
        : 0.02;
  } else {
    yTol = Math.max(10, Math.floor(imageMeta.height / 100));
  }

  const lines = groupIntoLines(ocrWords, yTol);
  let startIdx: number | null = null;
  for (let i = 0; i < lines.length; i++) {
    if (ROLE_KEYWORDS.test(lineText(lines[i]))) {
      startIdx = i;
      break;
    }
  }

  if (startIdx === null) {
    return {
      name: 'manufacturer_address',
      value: null,
      bbox: null,
      confidence: 0.0,
      evidence_spans: [],
    };
  }

  const blockLines: OCRWord[][] = [lines[startIdx]];
  const hasPin = blockLines[0].some((w) => pinRe.test(w.text));

  if (!hasPin) {
    const end = Math.min(startIdx + 7, lines.length);
    for (let j = startIdx + 1; j < end; j++) {
      if (lineIsAddress(lines[j], pinRe)) {
        blockLines.push(lines[j]);
        if (lines[j].some((w) => pinRe.test(w.text))) {
          break;
        }
      } else if (blockLines.length >= 2) {
        break;
      }
    }
  }

  const blockWords: OCRWord[] = [];
  for (const line of blockLines) {
    for (const w of line) {
      blockWords.push(w);
    }
  }

  const pinWord = blockWords.find((w) => pinRe.test(w.text));
  const pinBbox = pinWord ? pinWord.bbox : null;

  return {
    name: 'manufacturer_address',
    value: pinBbox ? blockLines.map(lineText).join(' ') : null,
    bbox: mergeBboxes(blockWords),
    confidence: pinBbox ? avgConfidence(blockWords) : 0.0,
    evidence_spans: pinBbox ? [pinBbox] : [],
  };
}
